File: node-server/src/notebook/cell-metadata.ts
/**
 * Cell Metadata Schema
 *
 * Defines which cell fields agents can modify and their types.
 * Must match lib/cellMetadata.ts on the frontend.
 *
 * Shared between:
 * - UI handler (useOperationHandler.ts)
 * - Headless handler (headless-handler.ts)
 * - Operation router validation
 */

export interface MetadataFieldSchema {
  type: 'string' | 'number' | 'boolean' | 'enum';
  values?: string[];
  agentMutable: boolean;
  default?: unknown;
  description?: string;
}

// Cell metadata schema - must match lib/cellMetadata.ts
export const CELL_METADATA_SCHEMA: Record<string, MetadataFieldSchema> = {
  id: {
    type: 'string',
    agentMutable: true,
    description: 'Unique cell identifier',
  },
  type: {
    type: 'enum',
    values: ['code', 'markdown'],
    agentMutable: true,
    description: 'Cell type: code for executable cells, markdown for documentation',
  },
  scrolled: {
    type: 'boolean',
    agentMutable: true,
    default: false,
    description: 'Whether cell output is collapsed (Jupyter standard)',
  },
  scrolledHeight: {
    type: 'number',
    agentMutable: true,
    description: 'Height in pixels when output is collapsed',
  },
};

export interface ValidationResult {
  valid: boolean;
  error?: string;
}

/**
 * Validate a metadata value against the schema.
 */
export function validateMetadataValue(key: string, value: unknown): ValidationResult {
  const schema = CELL_METADATA_SCHEMA[key];

  if (!schema) {
    const allowed = Object.keys(CELL_METADATA_SCHEMA).join(', ');
    return { valid: false, error: `Unknown field "${key}". Allowed: ${allowed}` };
  }

  if (!schema.agentMutable) {
    return { valid: false, error: `Field "${key}" is not modifiable` };
  }

  switch (schema.type) {
    case 'string':
      if (typeof value !== 'string') {
        return { valid: false, error: `Field "${key}" must be a string` };
      }
      break;
    case 'number':
      if (typeof value !== 'number') {
        return { valid: false, error: `Field "${key}" must be a number` };
      }
      break;
    case 'boolean':
      if (typeof value !== 'boolean') {
        return { valid: false, error: `Field "${key}" must be a boolean` };
      }
      break;
    case 'enum':
      if (!schema.values?.includes(value as string)) {
        return { valid: false, error: `Field "${key}" must be one of: ${schema.values?.join(', ')}` };
      }
      break;
  }

  return { valid: true };
}

File: deprecated/python-server/cell_metadata.py
"""
Cell Metadata Schema

Defines which cell fields agents can modify and their types.
Must match lib/cellMetadata.ts on the frontend.

Shared between:
- UI handler (useOperationHandler.ts)
- Headless handler (headless_handler.py)
- Operation router validation
"""

from typing import Dict, Any


# Cell metadata schema - must match lib/cellMetadata.ts
CELL_METADATA_SCHEMA = {
    'id': {'type': 'string', 'agent_mutable': True},
    'type': {'type': 'enum', 'values': ['code', 'markdown'], 'agent_mutable': True},
    'scrolled': {'type': 'boolean', 'agent_mutable': True, 'default': False},
    'scrolledHeight': {'type': 'number', 'agent_mutable': True},
}


def validate_metadata_value(key: str, value: Any) -> Dict[str, Any]:
    """Validate a metadata value against the schema. Returns {valid: bool, error?: str}"""
    schema = CELL_METADATA_SCHEMA.get(key)

    if schema is None:
        allowed = ', '.join(CELL_METADATA_SCHEMA.keys())
        return {'valid': False, 'error': f'Unknown field "{key}". Allowed: {allowed}'}

    if not schema.get('agent_mutable', False):
        return {'valid': False, 'error': f'Field "{key}" is not modifiable'}

    field_type = schema.get('type')
    if field_type == 'string':
        if not isinstance(value, str):
            return {'valid': False, 'error': f'Field "{key}" must be a string'}
    elif field_type == 'number':
        if not isinstance(value, (int, float)):
            return {'valid': False, 'error': f'Field "{key}" must be a number'}
    elif field_type == 'boolean':
        if not isinstance(value, bool):
            return {'valid': False, 'error': f'Field "{key}" must be a boolean'}
    elif field_type == 'enum':
        if value not in schema.get('values', []):
            return {'valid': False, 'error': f'Field "{key}" must be one of: {", ".join(schema["values"])}'}

    return {'valid': True}
